tensorflow_probability/python/distributions/empirical.py

```diff
@@ -0,0 +1,387 @@
+      samples: Numeric `Tensor` of shape [B1, ..., Bk, S, E]`, `k >= 0`.
+        Samples or batches of samples on which the distribution is based.
+        The first `k` dimensions index into a batch of independent distributions
+        and the last 2 dimension represents a multiset of vector-like samples
+        for each distribution.
+      validate_args: Python `bool`, default `False`. When `True` distribution
+        parameters are checked for validity despite possibly degrading runtime
+        performance. When `False` invalid inputs may silently render incorrect
+        outputs.
+      allow_nan_stats: Python `bool`, default `True`. When `True`, statistics
+        (e.g., mean, mode, variance) use the value "`NaN`" to indicate the
+        result is undefined. When `False`, an exception is raised if one or
+        more of the statistic's batch members are undefined.
+      name: Python `str` name prefixed to Ops created by this class.
+
+    Raises:
+      ValueError: if `samples` is a scalar.
+    """
+    super(VectorEmpirical, self).__init__(
+        samples,
+        event_ndims=1,
+        validate_args=validate_args,
+        allow_nan_stats=allow_nan_stats,
+        name=name)
+
+  def _mode(self):
+    def _get_mode(samples):
+      # TODO: Swith to tf.unique_with_counts_v2 when updated in TensorFlow API
+      y, idx, count = gen_array_ops.unique_with_counts_v2(samples, axis=[0])
+      mode = tf.gather(samples, tf.argmax(count))
+      return mode
+
+    event_size = tf.reshape(self.event_shape_tensor(), [])
+    samples = tf.reshape(self.samples, [-1, self.num_samples, event_size])
+    modes = tf.map_fn(_get_mode, samples)
+    modes_shape = tf.concat(
+        [self.batch_shape_tensor(), self.event_shape_tensor()],
+        axis=0)
+    return tf.reshape(modes, modes_shape)
+
+  def _sample_n(self, n, seed=None):
+    def _get_sample(samples):
+      indices = tf.random.uniform([n], maxval=self.num_samples,
+                                  dtype=tf.int32, seed=seed)
+      tensor = tf.gather(samples, indices, axis=-2)
+      return tensor
+
+    event_size = tf.reshape(self.event_shape_tensor(), [])
+    samples = tf.reshape(self.samples, [-1, self.num_samples, event_size])
+    draws = tf.map_fn(_get_sample, samples)
+    draws_shape = tf.concat(
+        [self.batch_shape_tensor(), self.event_shape_tensor()],
+        axis=0)
+    draws = tf.reshape(
+        tf.transpose(draws, perm=[1, 0, 2]),
+        tf.concat([[n], draws_shape], axis=0))
+    return draws
+
+  def _prob(self, event):
+    event = tf.convert_to_tensor(event, name='event', dtype=self.dtype)
+    if self.validate_args:
+      event = util.embed_check_integer_casting_closed(
+          event, target_dtype=tf.bool)
+    event, samples = _broadcast_event_and_samples(event, self.samples,
+                                                  event_ndims=1)
+
+    compare_samples_event = tf.cast(
+        tf.reduce_all(tf.equal(samples, event), axis=-1),
```

Comment:
you should be able to come up with a range for the axis parameter which would support higher & lower ranks.
I guess I'd prefer to see this exposed as Empirical(samples=..., event_ndims=...) so that we don't need VectorEmpirical special cased.

Reply:
That said, I'm sympathetic if you're ready to be done w/ this and I think we can take it as a TODO to generalize Empirical beyond scalar. We could still support scalar Empirical by defaulting event_ndims=0 and VectorEmpirical could subclass & override event_ndims=1.

Reply:
something like `compare_samples_event = tf.cast(tf.reduce_all(tf.equal(samples, event), axis=tf.range(-event_ndims-1, 1))...`

Reply:
I agree that it is great to have Empirical to support all event_ndims. Initially I was considering something similar to `Deterministic` and `VectorDeterminstic`, since the solution was not obvious to me. But thanks Brian, I just figured out a way to generalize it in one class. If it looks good to you, maybe we could also try that for Deterministic distribution?